### DISCUSSION TITLE
Support self bounded TypeVar for protected variable access

### DISCUSSION THREAD

**Discussion by aawilson** [microsoft/pyright#7708] [General]
Hello! I have a situation where I am trying to create a decorator on a class that wraps methods with code that uses a protected variable to manage context specific to the method. I'm using a `TypeVar` with a bound on my class to enforce that the callable passed to the decorator is a subtype of the class in question, but strict checking still throws `reportPrivateUsage`. I know I can disable this with a flag, but I was wondering if it was intended that TypeVar bounds checks not qualify for "protected" usage, or if I was perhaps misusing the syntax.

Playground [here](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4RQKZgPZSQBVgDcBDJDAVQGcS4MBeJKAVwwCgAzKbAWxCQhgYAOzggYPHHhABBYRAA0sgO4kYSEgCMANhiUBhEtu1bdB7MIDGZDMJvnuzJCL0gAIjEtIlAcVsZYSyV8EigSHgBlMAwgkAAVQQwANVC2NIAhbGZhABMMHIB5YX0TKioZOJB6eMSUqAAKACJM7LzC4tLyuMalTSzc%2BkaSkjKZRoBKNkrqhOi6pu7J-CqQELDI6Msm-Am0y06QYdGUNhAzkDyOEAB9a5F1W-qqDG0OJQAqULgqJUsLcgAHqh3J4kABtKhMJQiJAAXQ%2B7wA1qooN9xidzpiQM9XgA6a5gbjkLz5a5-YSApArcmU05Ys5UZjRBrjfF3YQPa71T6on4gd5IlFotJsS4XGK4Mi4a4caX7EZUa4kJVIJgwTRODBUep0s5IbDXHgYHjYGAALwwaEMxlMGDB%2Bgs1nIdnIYJauXyRSOXWhFOC8PEFPhbHGIAAtAA%2BQ5GEw6O0Oqw2F1291tL2dCr%2B31wjHnOxGlb6w3G00Wtn5jC3NKYsXKMJgaI5J4vDhoVOejoKzP8r5UND4XG9hHI3v93Ej3mhyOB1C6rGYJDMKDCfgGo0m80YZuvJQ4jj4wnYYnkHJk-4YIFgisBnnfYdCqiTOcLpcruskBv5NhAA)

**Comment by erictraut** [comment]
The `reportPrivateAccess` check is working as intended here. It's being used outside of the implementation of its class, which means it's not conforming to the notion of being "private". As you pointed out, you can suppress this check if you'd like.

**Reply by aawilson** [reply]
If you'll indulge me thinking through it a bit (pyright: ignore does practically solve the issue, and I'm happy with that if it it's the solution), and perhaps this stretches against the boundary of the semantics , but if the decorator implementation is limited to acting on subclasses of ClassA, isn't that semantically identical to defining a ClassB and using ClassA's protected methods within its implementation?

**Reply by erictraut** [reply]
Private and protected attributes are about information hiding and containment. The scope for a private or protected attribute is the class. Here you're trying to access the protected member from code that's outside the class, so you're violating the visibility rules of a protected attribute. The same argument would apply if you made the attribute private (i.e. renamed it with a double underscore) except the runtime would then enforce the private access, and you'd see a runtime exception when the decorator code executes.

I'm not sure what you mean by "semantically identical...". It's a function that is not part of the class implementation, and it's trying to access a private member in the class. If you exempt this function from this rule, why not exempt every function from this rule? At that point, the rule loses all of its meaning.

If you're looking for a way to avoid violating the protected attribute access, you could place the decorator implementation inside of `ClassA`. You can make it a `@staticmethod` like [this](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4RQKZgPZSQBVgDcBDJDAVQGcS4MBeJKAVwwCgAzKbAWxCQhgYAOzggYPHHhABhEgBt5JAEbyMAGlnZhAYzIZh%2BzQBEYOpJvwkoJHgGUwGHZoAqgjADVrbNgCFszMIAJhhBAPLCMkpUVACCLiD0IG6OXlAAFABE-oEh4ZHRcS6ZmsoBwfSZUSQxsZkAlGwJSSme1lnFjfiJIFY29o46WfgNPjqFsoWxKGwgcyAhHCAA%2BssiMEir6VQY8hyaAFTWcFSaOtrkAB6oIKbmANpUTJoiSAC6hwcA1gDux1T1GbzYEgHZ7AB0yzA3HI5lCy3OwiuSB6iORsxBcyozEcGXqkLWwg2WyOUBOn1%2B-0aGLmAAEnmQzDwMEgENggjSFhgliFzjYkLhlhxBeMalRliQJUgmDBlMxyFR0pzgQLlsyeNgYAAvDBoOSKFRqe4ybR6ciGcj3HLBUIRaq1FwvJGWD7iJEfTn1EAAWgAfLIFEpVBhjab9BaQ9a8napo7eq7Xm8gZiQIZmT1VerNTqCWmMKsfCmuUsfjYwI4gttdhw0FHbQUxfFDv80Phwf8KX8yVRW%2BDKd2vX63ahlSnMEhmFBhPxsGqMBrtRgq3tNGCOJDodhYeQggiLhhrvc867SeSQAdvl2TtSi%2BPJ9PSyRy6EfLzcGRBcKoAjChKpTK5QVHp7TicE335T8RV-SUyAA%2BUMCoNggA).

**Reply by aawilson** [reply]
All I mean by "Semantically identical" is in the difference between having a subclass definition access a superclass's protected attributes and writing a metamethod that defines a subclass method to access a superclass's protected attributes. In both cases, the protected attribute access happens away from the protected attribute's original definition, and using a bound (or a runtime issubclass check) seems to give the same meaning as putting the superclass in the MRO.

Either way, the current state of things is plenty satisfactory, and I again appreciate your indulgence in the thoughts above.
